### PR TITLE
Optimize Struct's accessors

### PR DESCRIPTION
**Pull request by ko1**
Improve performance of setter/getter of Struct with optimized path.

```yaml
  S = Struct.new(:a, :b)
  s = S.new(1, 2)
  h = {a: 1, b: 2}
  class C
    attr_accessor :a, :b
    def initialize; @a = 1; @b = 2; end
  end
  c = C.new
  require 'ostruct'
  o = OpenStruct.new(h)
benchmark:
  ivar: c.a; c.b
  Struct: s.a; s.b
  Hash: h[:a]; h[:b]
  ostruct: o.a; o.b
  ivar_set: c.a = c.b = nil
  Struct_set: s.a = s.b = nil
  Hash_set: h[:a] = h[:b] = nil
  ostruct_set: o.a = o.b = nil
```

```
master: ruby 3.1.0dev (2021-11-17T06:08:46Z master d23b3d9b7d) [x86_64-linux]
modified: ruby 3.1.0dev (2021-11-18T02:01:31Z opt_struct_aref 8f3366754b) [x86_64-linux]
Warming up --------------------------------------
                ivar    33.279M i/s -     33.309M times in 1.000920s (30.05ns/i, 113clocks/i)
              Struct    25.883M i/s -     25.904M times in 1.000804s (38.64ns/i, 146clocks/i)
                Hash    28.235M i/s -     28.318M times in 1.002938s (35.42ns/i, 134clocks/i)
             ostruct     7.813M i/s -      7.896M times in 1.010588s (127.99ns/i, 478clocks/i)
            ivar_set    29.306M i/s -     29.379M times in 1.002495s (34.12ns/i, 129clocks/i)
          Struct_set    17.276M i/s -     17.462M times in 1.010791s (57.88ns/i, 218clocks/i)
            Hash_set    16.574M i/s -     16.667M times in 1.005657s (60.34ns/i, 228clocks/i)
         ostruct_set     5.638M i/s -      5.692M times in 1.009573s (177.35ns/i, 673clocks/i)
Calculating -------------------------------------
                         master    modified
                ivar    60.246M     60.283M i/s -     99.836M times in 1.657155s 1.656131s
              Struct    39.569M     62.914M i/s -     77.649M times in 1.962351s 1.234199s
                Hash    45.743M     46.692M i/s -     84.706M times in 1.851763s 1.814144s
             ostruct     8.615M      8.757M i/s -     23.439M times in 2.720840s 2.676602s
            ivar_set    48.647M     45.752M i/s -     87.918M times in 1.807265s 1.921629s
          Struct_set    17.543M     46.291M i/s -     51.827M times in 2.954289s 1.119597s
            Hash_set    21.276M     21.404M i/s -     49.721M times in 2.336972s 2.322998s
         ostruct_set     6.177M      6.093M i/s -     16.915M times in 2.738419s 2.776225s

Comparison:
                             ivar
            modified:  60282947.4 i/s
              master:  60245692.0 i/s - 1.00x  slower

                           Struct
            modified:  62914175.1 i/s
              master:  39569174.9 i/s - 1.59x  slower

                             Hash
            modified:  46691778.1 i/s
              master:  45743222.5 i/s - 1.02x  slower

                          ostruct
            modified:   8756880.8 i/s
              master:   8614505.7 i/s - 1.02x  slower

                         ivar_set
              master:  48646710.3 i/s
            modified:  45751569.9 i/s - 1.06x  slower

                       Struct_set
            modified:  46291201.4 i/s
              master:  17543131.2 i/s - 2.64x  slower

                         Hash_set
            modified:  21403871.7 i/s
              master:  21275886.1 i/s - 1.01x  slower

                      ostruct_set
              master:   6177075.3 i/s
            modified:   6092956.8 i/s - 1.01x  slower
```

### CODE REVIEW

File: vm_eval.c [hunk 1]
```diff
@@ -163,6 +163,19 @@ vm_call0_cfunc(rb_execution_context_t *ec, struct rb_calling_info *calling, cons
...
 
+static void
+vm_call_check_arity(struct rb_calling_info *calling, int argc, const VALUE *argv)
+{
+    if (calling->kw_splat &&
+        calling->argc > 0 &&
+        RB_TYPE_P(argv[calling->argc-1], T_HASH) &&
+        RHASH_EMPTY_P(argv[calling->argc-1])) {
+        calling->argc--;
+    }
```
Comment (maximecb) [inline vm_eval.c:174]: It could be good to comment this to explain what this check is testing

**Comment by maximecb** [comment]
@XrXr @jhawthorn can you comment on whether this PR increases or decrease the complexity in YJIT, and how easy/hard it would be for us to support this change?

File: vm_eval.c (hunk 1 above)
Comment (ko1) [reply vm_eval.c:174]: @jeremyevans may know.

Comment (jeremyevans) [reply vm_eval.c:174]: This is testing for empty splatted keyword hashes.  If an empty splatted keyword hash are passed, it should be equivalent to not passing keywords at all.

**Comment by jhawthorn** [comment]
> can you comment on whether this PR increases or decrease the complexity in YJIT, and how easy/hard it would be for us to support this change?

I think it would very slightly increase complexity, since it's adding two new method types, but those method types should be pretty easy to implement. We should also be able to generate better code for these than we currently do (can now read from the memory location instead of calling the old builtin function).

**Comment by XrXr** [comment]
Yeah, we would need to add code to YJIT if we want to support this new method type. The actual logic is similar to getting an ivar, but a bit simpler because structs have a fixed layout. The relevant code is here:
https://github.com/ruby/ruby/blob/ab737b19197c63b84dad9944045a2fd2dc369264/internal/struct.h#L121-L144

So it's a flag test to find the buffer, and then reading at a fixed index.

**Comment by maximecb** [comment]
Since you guys don't seem concerned about the complexity, we can approve it. Would either of you want to implement the new struct access mechanism?

**Review by maximecb — APPROVED** [review]

**Review by tenderlove — APPROVED** [review]

**Comment by schneems** [comment]
After compiling with these changes I get a warning:

```
struct.c:267:1: warning: unused function 'struct_pos_num' [-Wunused-function]
struct_pos_num(VALUE s, VALUE idx)
^
1 warning generated.
```

**Comment by paracycle** [comment]
This PR seems to have changed the public API of `Struct` member setter methods.

Before:
```ruby
S = Struct.new(:foo)
S.instance_method(:foo=).parameters
#=> [[:req, :_]]
```

After:
```ruby
S = Struct.new(:foo)
S.instance_method(:foo=).parameters
#=> [[:rest]]
```

Is this intentional? I find it weird that a setter method looks like it is accepting zero or more arguments instead of a single required argument.

**Pull request merged**